mls-rs-crypto-webcrypto/src/ec/ecdh.rs

```diff
@@ -0,0 +1,239 @@
+
+        if private_key.is_ok() {
+            return Ok(private_key?);
+        }
+
+        // Last resort: try importing with a fake public key
+        der_private_key.set_fake_public_key(self.curve);
+
+        let private_key = der_private_key
+            .to_bytes(self.named_curve.clone())
+            .map_err(|e| CryptoError::DerError(e.to_string()))?;
+
+        Ok(private_key_type.import(crypto, &private_key).await?)
+    }
+
+    async fn generate(&self) -> Result<(HpkeSecretKey, HpkePublicKey), CryptoError> {
+        let (private, public) = generate(self.curve_name).await?;
+        Ok((private.into(), public.into()))
+    }
+
+    async fn to_public(&self, secret_key: &HpkeSecretKey) -> Result<HpkePublicKey, CryptoError> {
+        let crypto = get_crypto()?;
+        let private_key_type = KeyType::EcdhSecret(self.curve_name);
+
+        let der_secret_key = DerPrivateKey::from_raw(secret_key)
+            .map_err(|e| CryptoError::DerError(e.to_string()))?;
+
+        // Try providing a key without the public key
+        let bytes = der_secret_key
+            .to_bytes(self.named_curve.clone())
+            .map_err(|e| CryptoError::DerError(e.to_string()))?;
+
+        let private_key = private_key_type.import(&crypto, &bytes).await;
+
+        // If it worked, check if the public key has been computed
+        if let Ok(private_key) = &private_key {
+            let exported = crypto.export_key(private_key_type.format(), private_key)?;
+            let exported = JsFuture::from(exported).await?;
+            let exported = Uint8Array::new(&exported).to_vec();
+
+            let exported = DerPrivateKey::from_bytes(&exported)
+                .map_err(|e| CryptoError::DerError(e.to_string()))?;
+
+            if let Some(public_key) = exported.public_key {
+                return Ok(public_key.value.raw_bytes().to_vec().into());
+            }
+        }
+
+        Err(CryptoError::CouldNotComputePublicKey)
+    }
+}
+
+#[cfg_attr(not(mls_build_async), maybe_async::must_be_sync)]
+#[cfg_attr(mls_build_async, maybe_async::must_be_async(?Send))]
+impl DhType for Ecdh {
+    type Error = CryptoError;
+
+    async fn dh(
+        &self,
+        secret_key: &HpkeSecretKey,
+        public_key: &HpkePublicKey,
+    ) -> Result<Vec<u8>, Self::Error> {
+        self.dh(secret_key, public_key).await
+    }
+
+    async fn generate(&self) -> Result<(HpkeSecretKey, HpkePublicKey), Self::Error> {
+        self.generate().await
+    }
+
+    async fn to_public(&self, secret_key: &HpkeSecretKey) -> Result<HpkePublicKey, Self::Error> {
+        self.to_public(secret_key).await
+    }
+
+    fn bitmask_for_rejection_sampling(&self) -> Option<u8> {
+        self.curve.curve_bitmask()
+    }
+
+    fn secret_key_size(&self) -> usize {
+        self.curve.secret_key_size()
+    }
+
+    fn public_key_validate(&self, _key: &HpkePublicKey) -> Result<(), Self::Error> {
+        Ok(())
```

Comment:
Is it intended?

Reply:
Yeah, I'm not entirely sure what's the purpose of this function? Is this to make sure that when we encrypt to the key, we don't get encryption errors? If so, we don't call this often enough in the mls-rs crate. In any case, do we care about the scenario where someone sends us a correctly signed but malformed public key? (Note that they can already spread havoc by sending correctly signed but malformed ciphertexts, which we can't check.)